File: docs/ext/livecode.py
import re
import html

import requests
from docutils import nodes
from docutils.parsers.rst import Directive


LIVE_CODE_TEMPLATE = """<div class="livecode" livecode-default="{output}">{code}</div>"""

with open('../../single_include/pp2/pp2.h') as livecode_header:
    LIVE_CODE_HEADER = livecode_header.read()

ANSI_COLOR_CODE_PATTERN = re.compile(r'[\u001b\u009b][\[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-PRZcf-nqry=><]')


class LiveCodeDirective(Directive):
    has_content = True

    def compile(self, code):
        payload = {
            "source": (LIVE_CODE_HEADER + '\n' + code),
            "lang": "c++",
            "options": {
                "userArguments": "-E -P -ftrack-macro-expansion=0 -std=c++20",
                "compilerOptions": {
                      "skipAsm": False,
                      "executorRequest": False
                },
                "filters": {
                     "binary": False,
                     "commentOnly": True,
                     "demangle": True,
                     "directives": True,
                     "execute": False,
                     "intel": True,
                     "labels": True,
                     "libraryCode": False,
                     "trim": False
                },
                "tools": [],
                "libraries": []
            },
            "allowStoreCodeDebug": True
        }
        response = requests.post('https://godbolt.org/api/compiler/g121/compile', json=payload)
        response.encoding = 'utf-8'

        output = response.text
        output = output[output.find('\n')+1:]
        output = ANSI_COLOR_CODE_PATTERN.sub('', output)
        return output

    def quote(self, text):
        return html.escape(text, quote=True)

    def run(self):
        if self.has_content and len(self.content):
            code_input = '\n'.join(self.content)
            code_output = self.compile(code_input)
            code_block = LIVE_CODE_TEMPLATE.format(code=code_input, output=self.quote(code_output))
            return [nodes.raw('', code_block, format='html')]


def setup(app):
    app.add_directive("livecode", LiveCodeDirective)

    return {
        'version': '0.1',
        'parallel_read_safe': True,
        'parallel_write_safe': True,
    }

File: _static/scripts/livecode.js

$(document).ready(function(){
    function resizeFrames(layout, lhsHeight, rhsHeight) {
        const newHeight = Math.max(lhsHeight, rhsHeight);
        layout.updateSize("99.9%", newHeight);
    }

    function throttle(f, delay){
        var timer = null;
        return function(){
            var context = this, args = arguments;
            clearTimeout(timer);
            timer = window.setTimeout(function(){
                f.apply(context, args);
            },
            delay);
        };
    }

    function compile(code, callback) {
        const data = JSON.stringify({
            "source": (LIVECODE_HEADER + '\n' + code),
            "lang": "c++",
            "options": {
                "userArguments": "-E -P -ftrack-macro-expansion=0 -std=c++20",
                "compilerOptions": {
                      "skipAsm": false,
                      "executorRequest": false
                },
                "filters": {
                     "binary": false,
                     "commentOnly": true,
                     "demangle": true,
                     "directives": true,
                     "execute": false,
                     "intel": true,
                     "labels": true,
                     "libraryCode": false,
                     "trim": false
                },
                "tools": [],
                "libraries": []
            },
            "allowStoreCodeDebug": true
        });

        const requestOptions = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: data
        };

        fetch('https://godbolt.org/api/compiler/g121/compile', requestOptions)
            .then(response => response.text())
            .then(callback);
    }

    function attachHeightObserver(container, callback) {
        const ro = new ResizeObserver(entries => {
            const newHeight = Math.max.apply(null, entries.map(function(o){ return o.contentRect.height; }));
            callback(newHeight + 20);
        });
        ro.observe(container.getElement().find('.CodeMirror')[0]);
    }

    $(".livecode").each(function() {
        const codeInput = $(this).text().trim();
        const codeOutput = $(this).attr('livecode-default') || ''
        $(this).empty();

        var config = {
            settings: {
                hasHeaders: false,
                constrainDragToContainer: true,
                reorderEnabled: false,
                showPopoutIcon: false,
                showMaximiseIcon: false,
            },
            content: [{
                type: 'row',
                content: [{
                    type:'component',
                    componentName: 'livecode-code',
                    componentState: { text: codeInput },
                },
                {
                    type:'component',
                    componentName: 'livecode-output',
                    componentState: { text: codeOutput }
                }]
            }]
        };

        const myLayout = new window.GoldenLayout( config, $(this) );
        var liveCodeInput, liveCodeOutput;
        var lhsHeight = 0, rhsHeight = 0;

        myLayout.registerComponent( 'livecode-code', function( container, state ){
                liveCodeInput = CodeMirror(container.getElement()[0], {
                    value: state.text,
                    mode:  "stylus",
                    indentUnit: 4,
                    smartIndent: false,
                    styleActiveLine: true,
                    autoRefresh: true,
                    matchBrackets: true,
                    autoCloseBrackets: true,
                    viewportMargin: Infinity
                });

                liveCodeInput.on('change', throttle(function(instance, changeObj) {
                    compile(liveCodeInput.getValue(), function (data) {
                        liveCodeOutput.setValue(Colors.strip(data.replace(/.*/, "").slice(1)))
                    })
                }, 250));

                attachHeightObserver(container, function  (newHeight) {
                    lhsHeight = newHeight;
                    resizeFrames(myLayout, lhsHeight, rhsHeight);
                })
        });

        myLayout.registerComponent( 'livecode-output', function( container, state ){
            liveCodeOutput = CodeMirror(container.getElement()[0], {
                value: state.text,
                mode:  "null",
                autoRefresh: true,
                readOnly: true,
                viewportMargin: Infinity
            });

            attachHeightObserver(container, function  (newHeight) {
                rhsHeight = newHeight;
                resizeFrames(myLayout, lhsHeight, rhsHeight);
            })
        });

        myLayout.init();
        myLayout.updateSize("100%", 100);
    });
});